src/stm32/i2c.c

```diff
@@ -81,13 +82,20 @@ i2c_wait(I2C_TypeDef *i2c, uint32_t set, uint32_t clear, uint32_t timeout)
 }
 
 static void
-i2c_start(I2C_TypeDef *i2c, uint8_t addr, uint32_t timeout)
+i2c_start(I2C_TypeDef *i2c, uint8_t addr, uint8_t xfer_len,
+          uint32_t timeout)
 {
     i2c->CR1 = I2C_CR1_START | I2C_CR1_PE;
     i2c_wait(i2c, I2C_SR1_SB, 0, timeout);
     i2c->DR = addr;
+    if (addr & 0x01)
+        i2c->CR1 |= I2C_CR1_ACK;
     i2c_wait(i2c, I2C_SR1_ADDR, 0, timeout);
+    irq_disable();
```

Comment:
For intermediate code like this, it's better to use `irqstatus_t flag = irq_save(); ...; irq_restore(flag);` - that way the code does not enable interrupts if an i2c transfer was actually invoked while irqs were disabled.